kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ReverseFilesToColumnarBatchActionsIterable.java

```diff
@@ -0,0 +1,247 @@
+/*
+ * Copyright (2023) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.kernel.internal.replay;
+
+import io.delta.kernel.client.FileReadContext;
+import io.delta.kernel.client.JsonHandler;
+import io.delta.kernel.client.ParquetHandler;
+import io.delta.kernel.client.TableClient;
+import io.delta.kernel.data.FileDataReadResult;
+import io.delta.kernel.expressions.Literal;
+import io.delta.kernel.fs.FileStatus;
+import io.delta.kernel.internal.fs.Path;
+import io.delta.kernel.internal.lang.CloseableIterable;
+import io.delta.kernel.internal.util.InternalUtils;
+import io.delta.kernel.internal.util.Logging;
+import io.delta.kernel.types.StructType;
+import io.delta.kernel.utils.CloseableIterator;
+import io.delta.kernel.utils.Tuple2;
+import io.delta.kernel.utils.Utils;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * This class takes as input a list of delta files (.json, .checkpoint.parquet) and produces an
+ * iterator of (FileDataReadResult, isFromCheckpoint) tuples.
+ *
+ * Users can pass in a `readSchema` to select which actions they want to consume.
+ */
+public class ReverseFilesToColumnarBatchActionsIterable
+    implements CloseableIterable<Tuple2<FileDataReadResult, Boolean>>, Logging {
+
+    private final TableClient tableClient;
+    private final List<FileStatus> reverseSortedFiles;
+    private final StructType readSchema;
+
+    public ReverseFilesToColumnarBatchActionsIterable(
```

Comment:
Not sure what we are getting by making this an iterable. We are creating two instances of the iterable and calling iterator on it just once.

Also repeat the work of sorting the delta log commit files.